### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatoliy Bazko <abazko@redhat.com>

### CODE REVIEW

File: pkg/deploy/tls-secrets.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+			logrus.Infof("Waiting on job '%s' to be finished", CheTLSJobName)
+			if jobStatus.Err != nil {
+				logrus.Error(jobStatus.Err)
+			}
+
+			return TLSSecretsProvisioningStatus{ProvisioningStatus: jobStatus.ProvisioningStatus}
+		}
+
+		// cleanup job
+		if err = clusterAPI.Client.Delete(context.TODO(), jobStatus.Job); err != nil {
```
Comment (mmorhun) [inline pkg/deploy/tls-secrets.go:105]: You call delete after sync, looks very weird (I know how it works, but still).

File: pkg/deploy/tls-secrets.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+		cheTLSSecretsCreationJobImage := DefaultCheTLSSecretsCreationJobImage()
+		jobEnvVars := map[string]string{
+			"DOMAIN":                         checluster.Spec.K8s.IngressDomain,
+			"CHE_NAMESPACE":                  checluster.Namespace,
+			"CHE_SERVER_TLS_SECRET_NAME":     cheTLSSecretName,
+			"CHE_CA_CERTIFICATE_SECRET_NAME": CheTLSSelfSignedCertificateSecretName,
+		}
+		jobStatus := SyncJobToCluster(checluster, CheTLSJobName, cheTLSSecretsCreationJobImage, CheTLSJobServiceAccountName, jobEnvVars, clusterAPI)
+		if !jobStatus.Continue {
```
Comment (mmorhun) [inline pkg/deploy/tls-secrets.go:95]: We shouldn't delegate reconcile loop logic to helpers.

File: pkg/controller/che/che_controller.go [hunk 3]
```diff
@@ -268,13 +268,20 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 	// Handle Che TLS certificates on Kubernetes like infrastructures
 	if instance.Spec.Server.TlsSupport && instance.Spec.Server.SelfSignedCert && !isOpenShift {
 		// Ensure TLS configuration is correct
-		if err := CheckAndCorrectTLSConfiguration(instance, r); err != nil {
-			return reconcile.Result{}, err
+		if err := deploy.CheckAndUpdateTLSConfiguration(instance, clusterAPI); err != nil {
+			instance, _ = r.GetCR(request)
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/controller/che/che_controller.go:272]: When CR is updated then `metadata.resourceVersion` is automatically changed.
So getting the latest CR is required.

File: pkg/controller/che/che_controller.go [hunk 4]
```diff
@@ -268,13 +268,20 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+			return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/controller/che/che_controller.go:273]: `err != nil` 
something bad happened, we can't continue

File: pkg/controller/che/che_controller.go [hunk 5]
```diff
@@ -406,44 +413,105 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 	// create service accounts:
 	// che is the one which token is used to create workspace objects
 	// che-workspace is SA used by plugins like exec and terminal with limited privileges
-	cheServiceAccount := deploy.NewServiceAccount(instance, "che")
-	if err := r.CreateServiceAccount(instance, cheServiceAccount); err != nil {
-		return reconcile.Result{}, err
+	cheSAStatus := deploy.SyncServiceAccountToCluster(instance, "che", clusterAPI)
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/controller/che/che_controller.go:416]: I moved SA, role, rolebinding creation to a separate files, so these are side-effect of refactoring.

File: pkg/deploy/job.go [hunk 6]
```diff
@@ -30,50 +32,93 @@ import (
...
+	ProvisioningStatus
+	Job *batchv1.Job
+}
+
+func SyncJobToCluster(
+	checluster *orgv1.CheCluster,
+	name string,
+	image string,
+	serviceAccountName string,
+	env map[string]string,
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/deploy/job.go:62]: Let's think about it when it is needed.

File: pkg/controller/che/che_controller.go [hunk 7]
```diff
@@ -268,13 +268,20 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
...
+		if err := deploy.CheckAndUpdateTLSConfiguration(instance, clusterAPI); err != nil {
+			instance, _ = r.GetCR(request)
+			return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
 		}
+
 		// Create TLS secrets if needed
-		shouldReturn, reconsileResult, err := HandleCheTLSSecrets(instance, r)
-		if shouldReturn {
-			return reconsileResult, err
+		secretStatus := deploy.SyncTLSSecretsToCluster(instance, clusterAPI)
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/controller/che/che_controller.go:277]: I will restore to prev name

File: pkg/deploy/job.go [hunk 8]
```diff
@@ -30,50 +32,93 @@ import (
...
+	}
 }
 
-// GetSpecJob creates new job configuration by giben parameters.
-func GetSpecJob(checluster *orgv1.CheCluster, name string, namespace string, image string, serviceAccountName string, env map[string]string, backoffLimit int32, clusterAPI ClusterAPI) (*batchv1.Job, error) {
+// GetSpecJob creates new job configuration by given parameters.
+func getSpecJob(checluster *orgv1.CheCluster, name string, image string, serviceAccountName string, env map[string]string, clusterAPI ClusterAPI) (*batchv1.Job, error) {
 	labels := GetLabels(checluster, util.GetValue(checluster.Spec.Server.CheFlavor, DefaultCheFlavor))
 	labels["component"] = "che-create-tls-secret-job"
+	labels["job-name"] = name
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/deploy/job.go:115]: Agree. Not needed.

File: pkg/deploy/tls-secrets.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+	cheTLSSecretName := checluster.Spec.K8s.TlsSecretName
+
+	// ===== Check Che TLS certificate ===== //
+	cheTLSSecret := &corev1.Secret{}
+	err := clusterAPI.Client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: cheTLSSecretName}, cheTLSSecret)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading secret info
+			logrus.Errorf("Error getting Che TLS secert \"%s\": %v", cheTLSSecretName, err)
+			return TLSSecretsProvisioningStatus{ProvisioningStatus: ProvisioningStatus{Err: err}}
```
[the remark this answers is not in the document]
Comment (tolusha) [reply pkg/deploy/tls-secrets.go:48]: Kind of agree.

File: pkg/deploy/update.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+package deploy
+
+import (
+	"context"
+
+	orgv1 "github.com/eclipse/che-operator/pkg/apis/org/v1"
+	"github.com/sirupsen/logrus"
+)
+
+func UpdateCheCRSpec(instance *orgv1.CheCluster, updatedField string, value string, clusterAPI ClusterAPI) (err error) {
```
Comment (mmorhun) [inline pkg/deploy/update.go:21]: What if we need to update two fields at a time, say turn on TLS and self-signed-cert flag? Should we use an array, WDYT?

File: pkg/deploy/update.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	logrus.Infof("Updating %s CR with %s: %s", instance.Name, updatedField, value)
+	err = clusterAPI.Client.Update(context.TODO(), instance)
+	if err != nil {
+		logrus.Errorf("Failed to update %s CR: %s", instance.Name, err)
+		return err
+	}
+	logrus.Infof("Custom resource %s updated", instance.Name)
+	return nil
```
Comment (mmorhun) [inline pkg/deploy/update.go:29]: Does it make sense to return update CR? I mean requery it and return new object with new revision number. WDYT?

File: pkg/deploy/job.go [hunk 12]
```diff
@@ -22,58 +26,113 @@ import (
...
+			}
+			for _, env := range y {
+				yMap[env.Name] = env.Value
+			}
+			return reflect.DeepEqual(xMap, yMap)
+		}),
+	}
+)
+
+func SyncTlsJobToCluster(
```
Comment (mmorhun) [inline pkg/deploy/job.go:61]: Please remove this method.

File: pkg/deploy/job.go [hunk 13]
```diff
@@ -22,58 +26,113 @@ import (
...
-		logrus.Infof("Creating a new object: %s, name: %s", job.Kind, job.Name)
-		err = clusterAPI.Client.Create(context.TODO(), job)
+func SyncJobToCluster(
+	checluster *orgv1.CheCluster,
+	name string,
+	component string,
+	image string,
+	serviceAccountName string,
+	env map[string]string,
+	clusterAPI ClusterAPI) (*batchv1.Job, reconcile.Result, error) {
```
Comment (mmorhun) [inline pkg/deploy/job.go:78]: Isn't it a general helper? Please move reconcile decision into controller package.

File: pkg/deploy/update.go (hunk 10 above)
Comment (tolusha) [reply pkg/deploy/update.go:21]: We have to invoke `UpdateCheCRSpec`  twice

File: pkg/deploy/update.go (hunk 11 above)
Comment (tolusha) [reply pkg/deploy/update.go:29]: I don't want to touch it. it is used all over the project.

File: pkg/deploy/job.go (hunk 12 above)
Comment (tolusha) [reply pkg/deploy/job.go]: done

File: pkg/deploy/tls-secrets.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package deploy
```
Comment (mmorhun) [inline pkg/deploy/tls-secrets.go:12]: I think it is controller logic

File: pkg/deploy/tls-secrets.go [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+		cheTLSSecretsCreationJobImage := DefaultCheTLSSecretsCreationJobImage()
+		jobEnvVars := map[string]string{
+			"DOMAIN":                         checluster.Spec.K8s.IngressDomain,
+			"CHE_NAMESPACE":                  checluster.Namespace,
+			"CHE_SERVER_TLS_SECRET_NAME":     cheTLSSecretName,
+			"CHE_CA_CERTIFICATE_SECRET_NAME": CheTLSSelfSignedCertificateSecretName,
+		}
+		job, err := SyncJobToCluster(checluster, CheTLSJobName, CheTlsJobComponentName, cheTLSSecretsCreationJobImage, CheTLSJobServiceAccountName, jobEnvVars, clusterAPI)
+		if err != nil || job == nil || job.Status.Succeeded == 0 {
+			logrus.Infof("Waiting on job '%s' to be finished", CheTLSJobName)
```
Comment (mmorhun) [inline pkg/deploy/tls-secrets.go:100]: I don't think we wait job when it is `nil` or an error happened...

File: pkg/deploy/tls-secrets.go [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+
+		cheTLSSecretsCreationJobImage := DefaultCheTLSSecretsCreationJobImage()
+		jobEnvVars := map[string]string{
+			"DOMAIN":                         checluster.Spec.K8s.IngressDomain,
+			"CHE_NAMESPACE":                  checluster.Namespace,
+			"CHE_SERVER_TLS_SECRET_NAME":     cheTLSSecretName,
+			"CHE_CA_CERTIFICATE_SECRET_NAME": CheTLSSelfSignedCertificateSecretName,
+		}
+		job, err := SyncJobToCluster(checluster, CheTLSJobName, CheTlsJobComponentName, cheTLSSecretsCreationJobImage, CheTLSJobServiceAccountName, jobEnvVars, clusterAPI)
+		if err != nil || job == nil || job.Status.Succeeded == 0 {
```
Comment (mmorhun) [inline pkg/deploy/tls-secrets.go:99]: If the job fails, this logic will never restart (recreate) it. Am I right?

File: pkg/deploy/tls-secrets.go (hunk 14 above)
Comment (tolusha) [reply pkg/deploy/tls-secrets.go:12]: It is just a package name :)

Comment (tolusha) [reply pkg/deploy/tls-secrets.go:12]: Everything is controller.

Comment (mmorhun) [reply pkg/deploy/tls-secrets.go:12]: Why do we need packages then :)

File: pkg/deploy/tls-secrets.go (hunk 15 above)
Comment (tolusha) [reply pkg/deploy/tls-secrets.go]: ok

File: pkg/deploy/tls-secrets.go (hunk 14 above)
Comment (tolusha) [reply pkg/deploy/tls-secrets.go:12]: In controller package we methods for ReconcileChe structure only.
So, it make sense to move tls-secrets.go to deploy package.
Let's keep as is

**Pull request merged**